src/main/scala/io/citrine/lolo/trees/multitask/MultiTaskTrainingNode.scala

```diff
@@ -2,171 +2,207 @@ package io.citrine.lolo.trees.multitask
+            trainingWeight = trainingData.map(_._3).sum
           )
         } else {
-          if (!label.isInstanceOf[Char]) throw new IllegalArgumentException("Training data wasn't double or char")
           new InternalModelNode[PredictionResult[Char]](
-            split,
-            theLeftChild.getNode(index).asInstanceOf[ModelNode[PredictionResult[Char]]],
-            theRightChild.getNode(index).asInstanceOf[ModelNode[PredictionResult[Char]]],
-            outputDimension = 0, // Don't support multitask SHAP at this time.
-            trainingWeight = reducedData.length.toDouble
+            split = split,
+            left = leftNodeOption.get.getNode(index).asInstanceOf[ModelNode[PredictionResult[Char]]],
+            right = rightNodeOption.get.getNode(index).asInstanceOf[ModelNode[PredictionResult[Char]]],
+            outputDimension = 0,
+            trainingWeight = trainingData.map(_._3).sum
           )
         }
-      case (Some(theLeftChild), _) if left.nonEmpty =>
-        theLeftChild.getNode(index)
-      case (_, Some(theRightChild)) if right.nonEmpty =>
-        theRightChild.getNode(index)
-      case (_, _) =>
-        // TODO (PLA-10415): get the rng in here somehow (right now it's instantiating a random rng)
-        if (label.isInstanceOf[Double]) {
-          new RegressionTrainingLeaf(
-            reducedData.asInstanceOf[Seq[(Vector[AnyVal], Double, Double)]],
-            GuessTheMeanLearner(),
-            1,
-            rng = Random()
-          ).getNode()
+      case Inaccessible() => throw new RuntimeException(s"No valid training data present for label $index")
+    }
+  }
+}
+
+object MultiTaskTrainingNode {
+
+  /**
+    * Build a node by computing splits and recursively building the child nodes.
+    *
+    * @param trainingData on which to build the split
+    * @param numFeatures to considering when determining the split
+    * @param remainingDepth the maximum number of splits left
+    * @param maxDepth maximum number of splits (used to compute depth of this node)
+    * @param minInstances minimum training instances per node
+    * @param splitter determines the best split
+    * @param rng random number generator, for reproducibility
+    * @return
+    */
+  def build(
+      trainingData: Seq[(Vector[AnyVal], Array[AnyVal], Double)],
+      numFeatures: Int,
+      remainingDepth: Int,
+      maxDepth: Int,
+      minInstances: Int,
+      splitter: MultiTaskSplitter,
+      rng: Random = Random()
+  ): MultiTaskTrainingNode = {
+    val sufficientData = trainingData.size >= 2 * minInstances &&
+      remainingDepth > 0 &&
+      trainingData.exists(row => !row._2.sameElements(trainingData.head._2))
+    val (split: Split, deltaImpurity: Double) = if (sufficientData) {
+      splitter.getBestSplit(
+        trainingData,
+        numFeatures,
+        minInstances,
+        rng
+      )
+    } else {
+      (NoSplit(), 0.0)
+    }
+
+    val exampleRow = trainingData.head
```

Comment:
I don't think it works well as its own method because you have to pass so much context (about 6-7 arguments) that it ends up being pretty bulky. I added comments to segment the logic.